File: packages/javascript/lib/node/file/tailFile.ts
import { open } from 'fs/promises';

const CHUNK_SIZE = 64 * 1024;

// Node's readline (`crlfDelay: Infinity`) breaks a line on '\n', '\r\n' and a
// lone '\r'.
const LINE_BREAK_REGEX = /\r\n|\n|\r/;

function splitLines(text: string): string[] {
	const lines = text.split(LINE_BREAK_REGEX);

	// A break at the end of the text closes the last line, it does not open an
	// empty one. This is what readline does.
	if (lines.length > 1 && lines[lines.length - 1] === '') {
		lines.pop();
	}

	return lines;
}

export async function tailFile(filePath: string, length = 1): Promise<string | null> {
	if (length <= 0) {
		return null;
	}

	const handle = await open(filePath, 'r');

	try {
		const { size } = await handle.stat();

		if (size === 0) {
			return null;
		}

		// Read backwards from the end of the file so the work follows the number
		// of lines asked for rather than the size of the file. Reading forwards
		// means a 10 GB log is streamed in full to answer for its last line.
		const chunks: Buffer[] = [];
		let position = size;
		let lines: string[] = [];

		while (position > 0) {
			const readLength = Math.min(CHUNK_SIZE, position);
			position -= readLength;

			const chunk = Buffer.alloc(readLength);

			await handle.read(chunk, 0, readLength, position);
			chunks.unshift(chunk);

			lines = splitLines(Buffer.concat(chunks).toString('utf-8'));

			// One line more than asked for means the first line held in the
			// buffer is complete garbage - it starts wherever the chunk boundary
			// fell, possibly mid-character - but it is also about to be dropped.
			if (lines.length > length) {
				break;
			}
		}

		const buffer = lines.slice(-length);

		// The last line of newline characters is ignored.
		if (buffer.length > 0 && buffer[buffer.length - 1] === '') {
			buffer.pop();
		}

		if (buffer.length === 0) {
			return null;
		}

		return buffer.join('\n');
	} finally {
		await handle.close();
	}
}

File: packages/python/qsu/file/tailFile.py
import os
import re

_CHUNK_SIZE = 64 * 1024

# Node's readline (createInterface with crlfDelay: Infinity) breaks a line on
# '\n', '\r\n' and a lone '\r'.
_LINE_BREAK_REGEX = re.compile(r'\r\n|\n|\r')


def _splitLines(text: str):
	lines = _LINE_BREAK_REGEX.split(text)

	# A break at the end of the text closes the last line, it does not open an
	# empty one.
	if len(lines) > 1 and lines[-1] == '':
		lines.pop()

	return lines


def tailFile(filePath: str, length: int = 1):
	if length <= 0:
		return None

	lines = []

	with open(filePath, 'rb') as stream:
		stream.seek(0, os.SEEK_END)
		size = stream.tell()

		if size == 0:
			return None

		# Read backwards from the end of the file so the work follows the number
		# of lines asked for rather than the size of the file. Reading forwards
		# means a 10 GB log is streamed in full to answer for its last line.
		position = size
		chunks: list = []

		while position > 0:
			readLength = min(_CHUNK_SIZE, position)
			position -= readLength

			stream.seek(position)
			chunks.insert(0, stream.read(readLength))

			lines = _splitLines(b''.join(chunks).decode('utf-8', errors='replace'))

			# One line more than asked for means the first line held in the
			# buffer is incomplete - it starts wherever the chunk boundary fell,
			# possibly mid-character - but it is also about to be dropped.
			if len(lines) > length:
				break

	buffer = lines[-length:]

	# The last line of newline characters is ignored.
	if len(buffer) > 0 and buffer[-1] == '':
		buffer.pop()

	if len(buffer) == 0:
		return None

	return '\n'.join(buffer)
